### PR TITLE
Initial DDP implementation 

### PR DESCRIPTION
**Pull request by celiaberon**
Initial implementation of data distributed processing. Tested only on single node configurations. Includes:
- Initialize processes within experiment based on detected world size, set up main process as rank 0.
- [AlignmentStats only]: Wrap all networks in torch DDP, forcing referencing to any attribute written for AlignmentNetwork class. 
- Training loop aggregation of custom metrics - all_reduce and all_gather, depending on data type. All processes sync up measurements to allow for downstream processing.
- for targeted dropout, sum up all dropout inference passes across process and take mean over collective batches with all_reduce(sum)
- measure eigenfeatures gets sample size standardization, gather all features 
	- Note: current assumption that class names will match between processes
- eigenfeature_dropout uses same approach as with targeted dropout, but taking the appropriate eigenfeatures to match with dataset (therefore different inputs for each process...to consider: maybe we don't want the all_reduce(), but instead separate scores corresponding to each set of inputs

To do:
- Optimize moving metrics between devices if it's causing substantial delays. All distributed agg functions need data on GPU, whereas any plotting needs it back on CPU.
- With DDP, effective batch size for training is batch_size x world_size. However, some metrics (e.g. alignment, eigenfeatures) are just calculated on batch_size (such that we get world_size measurements of each metric). Decide proper handling - can gather and then split, with plot replicates (current eigenfeature implementation), or maintain separate streams through experiment on each process. Alignment tracking into parallel streams is incomplete.
- Compatibility testing across configurations: Multi node, single GPU, CPU
- Streamlined workflow for syncing new metrics (e.g. delta_alignment, expanding implementation to other experiment workflows)

Other new features:
- Unique checkpointing throughout experiment
- Utility functions replicating arbitrary nested list structures

### CODE REVIEW

File: networkAlignmentAnalysis/train.py [hunk 1]
```diff
@@ -53,22 +55,37 @@ def train(nets, optimizers, dataset, **parameters):
     # --- create results dictionary if not provided and handle checkpoint info ---
     results = parameters.get("results", False)
     num_complete = parameters.get("num_complete", 0)
-    save_ckpt, freq_ckpt, path_ckpt, dev = parameters.get("save_checkpoints", (False, 1, "", ""))
+    save_ckpt = parameters.get("save_ckpt", False)
```
Comment (landoskape) [inline networkAlignmentAnalysis/train.py:58]: ah this is good to break this up

**Comment by landoskape** [comment]
Amazing work, this is a lot of really excellent code. Thanks Celia!
